hack/boilerplate/boilerplate.go

```diff
@@ -0,0 +1,158 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+)
+
+var (
+	skippedPaths   = regexp.MustCompile(`Godeps|third_party|_gopath|_output|\.git|cluster/env.sh|vendor|test/e2e/generated/bindata.go|site/themes/docsy`)
+	rootdir        *string
+	boilerplatedir *string
+)
+
+func init() {
+	cwd, _ := os.Getwd()
+	boilerplatedir = flag.String("boilerplate-dir", cwd, "Boilerplate directory for boilerplate files")
+	cwd += "/../../"
+	rootdir = flag.String("rootdir", filepath.Dir(cwd), "Root directory to examine")
+}
+
+func main() {
+	flag.Parse()
+	refs := boilerplateRefs(*boilerplatedir)
+	if len(refs) == 0 {
+		log.Fatal("no references in ", *boilerplatedir)
+	}
+	files := filesToCheck(*rootdir, refs)
+	for _, file := range files {
+		if !filePasses(file, refs[fileExtension(file)]) {
+			fmt.Println(file)
+		}
+	}
+
+}
+
+/*
+This function is to populate the refs variable with the
+different boilerplate/template for different extension.
+*/
+func boilerplateRefs(dir string) map[string][]byte {
+	refs := make(map[string][]byte)
+	files, _ := filepath.Glob(dir + "/*.txt")
+	for _, filename := range files {
+		re := regexp.MustCompile(`\.txt`)
+		extension := strings.ToLower(fileExtension(re.ReplaceAllString(filename, "")))
+		data, err := ioutil.ReadFile(filename)
+		if err != nil {
+			log.Fatal(err)
+		}
+		re = regexp.MustCompile(`\r`)
+		refs[extension] = re.ReplaceAll(data, nil)
+	}
+	return refs
+}
+
+/*
+Function to check whether the processed file
+is valid.
+Returning false means that the file does not the
+proper boilerplate template
+*/
+func filePasses(filename string, ref []byte) bool {
+	var re *regexp.Regexp
+	data, err := ioutil.ReadFile(filename)
+	if err != nil {
+		log.Fatal(err)
+	}
+	re = regexp.MustCompile(`\r`)
+	data = re.ReplaceAll(data, nil)
+
+	extension := fileExtension(filename)
+
+	// remove build tags from the top of Go files
+	if extension == "go" {
+		re = regexp.MustCompile(`(?m)^(// \+build.*\n)+\n`)
+		data = re.ReplaceAll(data, nil)
+	}
+
+	// remove shebang from the top of shell files
+	if extension == "sh" {
+		re = regexp.MustCompile(`(?m)^(#!.*\n)\n*`)
+		data = re.ReplaceAll(data, nil)
+	}
+
+	// if our test file is smaller than the reference it surely fails!
+	if len(data) < len(ref) {
+		return false
+	}
+
+	data = data[:len(ref)]
+
+	// Search for "Copyright YEAR" which exists in the boilerplate, but shouldn't in the real thing
+	re = regexp.MustCompile(`Copyright YEAR`)
```

Comment:
This check should be only performed once when loading boilerplate text - in the function currently called `boilerplateRefs`.

Reply:
We're not checking the boilerplate here, but check every file if it contains the template YEAR instead of the actual number